File: cpp/structure/tree/binary_tree/rebuild/RebuildOfInAndPostOrder.cpp
#include <iostream>
#include <vector>
using namespace std;

struct TreeNode {
    int value;
    TreeNode *left;
    TreeNode *right;

    TreeNode(int value) : value(value), left(nullptr), right(nullptr) {}
};

class RebuildOfInAndPostOrder {
public:
    TreeNode *rebuild(const vector<int> &inOrder, const vector<int> &postOrder) {
        int pivot = postOrder[postOrder.size() - 1];
        TreeNode *parent = new TreeNode(pivot);
        if (1 == postOrder.size()) {
            return parent;
        }

        int index = 0;
        while (index < (int) postOrder.size() && pivot != inOrder[index]) {
            ++index;
        }

        if (index > 0) {
            parent->left = rebuild(vector<int>(inOrder.begin(), inOrder.begin() + index), vector<int>(postOrder.begin(), postOrder.begin() + index));
        }
        if (index < (int) postOrder.size() - 1) {
            parent->right = rebuild(vector<int>(inOrder.begin() + index + 1, inOrder.end()), vector<int>(postOrder.begin() + index, postOrder.end() - 1));
        }
        return parent;
    }
};

File: py/structure/tree/binary_tree/rebuild/RebuildOfInAndPostOrder.py
#!/usr/bin/env python
# -*- coding:gbk -*-

from typing import List

class TreeNode:
    def __init__(self, value: int):
        self.value, self.left, self.right = value, None, None

class RebuildOfInAndPostOrder:
    def rebuild(self, inOrder: List[int], postOrder: List[int]) -> TreeNode:
        pivot = postOrder[-1]
        parent = TreeNode(pivot)
        if 1 == len(postOrder):
            return parent

        index = 0
        while index < len(postOrder) and pivot != inOrder[index]:
            index += 1

        if index > 0:
            parent.left = self.rebuild(inOrder[ : index], postOrder[ : index])
        if index < len(postOrder) - 1:
            parent.right = self.rebuild(inOrder[index + 1 : ], postOrder[index : -1])
        return parent
